app/IATI/Services/Publisher/PublisherService.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\IATI\Services\Publisher;
+
+use App\IATI\Services\Workflow\RegistryApiHandler;
+use App\IATI\Traits\RegistryApiInvoker;
+use GuzzleHttp\Exception\ClientException;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Facades\Storage;
+
+/**
+ * Class PublisherService.
+ */
+class PublisherService extends RegistryApiHandler
+{
+    use RegistryApiInvoker;
+
+    /**
+     * @var Model|null
+     */
+    protected $activityPublished = null;
+
+    /**
+     * Publishes the activity xml file to the IATI registry.
+     *
+     * @param $registryInfo
+     * @param $activityPublished
+     * @param $organization
+     *
+     * @return void
+     */
+    public function publishFile($registryInfo, $activityPublished, $organization): void
+    {
+        $this->setFile($activityPublished);
+        $this->init(env('IATI_API_ENDPOINT'), Arr::get($registryInfo, 'api_token', ''))
+             ->setPublisher(Arr::get($registryInfo, 'publisher_id', ''));
+        $this->searchForPublisher($this->publisherId);
+        $this->publishToRegistry($organization, $activityPublished->filename);
+    }
+
+    /**
+     * Set the file attribute.
+     *
+     * @param $activityPublished
+     */
+    protected function setFile($activityPublished)
+    {
+        $this->activityPublished = $activityPublished;
+    }
+
+    /**
+     * Publish File to the IATI Registry.
+     *
+     * @param $organization
+     * @param $filename
+     *
+     * @return void
+     */
+    protected function publishToRegistry($organization, $filename): void
+    {
+        $data = $this->generatePayload($organization, $filename);
+        $packageId = $this->extractPackage($filename);
+
+        if ($this->isPackageAvailable($packageId, $this->apiKey)) {
+            $this->client->package_update($data);
+        } else {
+            $this->client->package_create($data);
+        }
+
+        $this->updateStatus();
+    }
+
+    /**
+     * Returns the request header payload while publishing any files to the IATI Registry.
+     *
+     * @param      $organization
+     * @param      $filename
+     * @param      $publishingType
+     * @param null $publishedFile
+     *
+     * @return string
+     */
+    protected function generatePayload($organization, $filename): string
+    {
+        $code = $this->getCode($filename);
+        $key = $this->getKey($code);
+        $fileType = $this->getFileType($code);
+        $title = $this->extractTitle($organization, $fileType);
+
+        return $this->formatHeaders($this->extractPackage($filename), $organization, $this->activityPublished, $key, $fileType, $title);
+    }
+
+    /**
+     * Get the data type or country code/region code from the filename.
+     *
+     * @param $filename
+     *
+     * @return string
+     */
+    protected function getCode($filename): string
+    {
+        $filename = str_replace('.xml', '', $filename);
```

Comment:
return substr(str_replace('.xml', '', $filename), strlen($this->publisherId) + 1);